File: qt6-widgets/examples/cpp/ch03_custom_signal.cpp
// ch03_custom_signal.py と同じことを C++ で書いたもの。
//
// 自分でシグナルを作るときだけ、C++ には Python にない手続きが要ります。
//   ・クラスに Q_OBJECT マクロを書く
//   ・そのクラスを moc（メタオブジェクトコンパイラ）に通す
//
// moc の実行は CMake の CMAKE_AUTOMOC が自動でやってくれるので、
// 実際に自分で意識するのは Q_OBJECT を書き忘れないことだけです。

#include <QApplication>
#include <QLabel>
#include <QLineEdit>
#include <QPushButton>
#include <QVBoxLayout>
#include <QWidget>

class NameForm : public QWidget
{
    Q_OBJECT   // ★ これを書き忘れると、シグナルもスロットも動かない

public:
    NameForm()
    {
        setWindowTitle("自作シグナル（C++）");
        resize(380, 170);

        m_edit = new QLineEdit;
        m_edit->setPlaceholderText("名前を入れて［あいさつ］を押す");

        QPushButton *button = new QPushButton("あいさつ");
        m_result = new QLabel("―");

        QVBoxLayout *layout = new QVBoxLayout(this);
        layout->addWidget(m_edit);
        layout->addWidget(button);
        layout->addWidget(m_result);

        connect(button, &QPushButton::clicked, this, &NameForm::emitSubmitted);
        connect(this, &NameForm::submitted, this, &NameForm::greet);
    }

signals:
    // Python: submitted = Signal(str)
    void submitted(const QString &name);

private slots:
    void emitSubmitted()
    {
        emit submitted(m_edit->text());
    }

    // Python: @Slot(str) def greet(self, name)
    void greet(const QString &name)
    {
        m_result->setText(QString("こんにちは、%1 さん！")
                              .arg(name.isEmpty() ? QString("ななし") : name));
    }

private:
    QLineEdit *m_edit;
    QLabel *m_result;
};

int main(int argc, char *argv[])
{
    QApplication app(argc, argv);

    NameForm form;
    form.show();

    return app.exec();
}

// 1 ファイルにクラスを書いた場合は、moc の生成結果をここで取り込む。
// ヘッダとソースを分けるふつうの書き方なら、この行は要りません。
#include "ch03_custom_signal.moc"

File: qt6-widgets/examples/ch03_custom_signal.py
"""自分でシグナルを作る — Signal と @Slot。

自作のシグナルは「クラス変数」として宣言する。
インスタンス変数（self.xxx = Signal()）にしても動かないので注意。

実行:  python ch03_custom_signal.py
"""

import sys

from PySide6.QtCore import Signal, Slot
from PySide6.QtWidgets import (QApplication, QLabel, QLineEdit, QPushButton,
                               QVBoxLayout, QWidget)


class NameForm(QWidget):
    # ★ クラス直下で宣言する。str は「このシグナルが文字列を1つ運ぶ」という意味。
    submitted = Signal(str)

    def __init__(self):
        super().__init__()
        self.setWindowTitle("自作シグナル")
        self.resize(380, 170)

        self.edit = QLineEdit()
        self.edit.setObjectName("nameEdit")
        self.edit.setPlaceholderText("名前を入れて［あいさつ］を押す")

        button = QPushButton("あいさつ")
        button.setObjectName("greetButton")
        button.clicked.connect(self._emit_submitted)

        self.result = QLabel("―")

        layout = QVBoxLayout(self)
        layout.addWidget(self.edit)
        layout.addWidget(button)
        layout.addWidget(self.result)

        # 自分のシグナルを、自分のスロットにつなぐ。
        self.submitted.connect(self.greet)

    def _emit_submitted(self):
        # emit() でシグナルを「発射」する。
        self.submitted.emit(self.edit.text())

    @Slot(str)
    def greet(self, name: str):
        """@Slot を付けると Qt 側に型が伝わり、わずかに速く・安全になる。"""
        self.result.setText(f"こんにちは、{name or 'ななし'} さん！")


app = QApplication(sys.argv)
form = NameForm()
form.show()
sys.exit(app.exec())
